### PR TITLE
reflect: treat proxy types correctly when serializing

### LINKED ISSUE #12001: Serialization of `Name` component may not be working correctly
## Bevy version

0.13

## What you did

Modified the `scene` example to include `Name::new("TEST")`.

Weirdly, I also had to `.register_type::<Name>()` in the example's `App`, despite it being registered in [bevy_core](https://github.com/bevyengine/bevy/blob/v0.13.0/crates/bevy_core/src/lib.rs#L42).

<details>
<summary>
Diff
</summary>

```rust
diff --git a/examples/scene/scene.rs b/examples/scene/scene.rs
index dac36e96d..3acb19f11 100644
--- a/examples/scene/scene.rs
+++ b/examples/scene/scene.rs
@@ -8,6 +8,7 @@ fn main() {
         .register_type::<ComponentA>()
         .register_type::<ComponentB>()
         .register_type::<ResourceA>()
+        .register_type::<Name>()
         .add_systems(
             Startup,
             (save_scene_system, load_scene_system, infotext_system),
@@ -114,6 +115,7 @@ fn save_scene_system(world: &mut World) {
         component_b,
         ComponentA { x: 1.0, y: 2.0 },
         Transform::IDENTITY,
+        Name::new("TEST"),
     ));
     scene_world.spawn(ComponentA { x: 3.0, y: 4.0 });
     scene_world.insert_resource(ResourceA { score: 1 });
```
</details>

`cargo run --example scene`

And observe the serialized scene in the logged output.

## What went wrong

`Name` is serialized as:

```plain
"bevy_core::name::Name": (
  hash: 778249913144375361,
  name: "TEST",
),
```

Which is contrary to this comment on name:
https://github.com/bevyengine/bevy/blob/abb8c353f49a6fe9e039e82adbe1040488ad910a/crates/bevy_core/src/name.rs#L23-L29

And the stated goal of #11447.

And the migration guide text which suggests that `Name` will be serialized as a plain string.

## Additional information

Scenes will properly **de**serialize without `hash`, but only as a struct with a `name` field. Not as a plain string.

See also: #1037 where we removed the migration guide entry related to #11447.

### PR DESCRIPTION
**Pull request by soqb**
# Objective

- Fixes #12001.
- Note this PR doesn't change any feature flags, however flaky the issue revealed they are.

## Solution

- Use `FromReflect` to convert proxy types to concrete ones in `ReflectSerialize::get_serializable`.
- Use `get_represented_type_info() -> type_id()` to get the correct type id to interact with the registry in `bevy_reflect::serde::ser::get_serializable`.

---

## Changelog

- Registering `ReflectSerialize` now imposes additional `FromReflect` and `TypePath` bounds.

## Migration Guide

- If `ReflectSerialize` is registered on a type, but `TypePath` or `FromReflect` implementations are omitted (perhaps by `#[reflect(type_path = false)` or `#[reflect(from_reflect = false)]`), the traits must now be implemented.

### CODE REVIEW

**Comment by rparrett** [comment]
Not a proper review, but Is this the expected output?

```rust
// with `serialize`
"bevy_core::name::Name": "joe",

// without
"bevy_core::name::Name": (
  hash: 1651241057899607384,
  name: "joe",
),
```

If so, does this comment on `Name` need to be adjusted?
```rust
hash: u64, // Won't be serialized (see: `bevy_core::serde` module) 
```
Or do we need a followup issue to address `hash` being serialized when `serialize` isn't enabled?

**Comment by soqb** [comment]
yeah i chose not to do anything to the feature flags to make this fix as simple as possible - i don't think it's trivial (and it's probably not uncontroversial) to remove all the feature flags for serialisation throughout bevy.

File: crates/bevy_reflect/src/type_registry.rs [hunk 1]
```diff
@@ -435,14 +435,22 @@ pub struct ReflectSerialize {
     get_serializable: for<'a> fn(value: &'a dyn Reflect) -> Serializable,
 }
 
-impl<T: Reflect + erased_serde::Serialize> FromType<T> for ReflectSerialize {
+impl<T: TypePath + Reflect + FromReflect + erased_serde::Serialize> FromType<T>
```
Comment (MrGVSV) [inline crates/bevy_reflect/src/type_registry.rs:438]: Nit:
```suggestion
impl<T: TypePath + FromReflect + erased_serde::Serialize> FromType<T>
```

**Review by MrGVSV — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@soqb can you run `cargo fmt` on this for me?

**Pull request merged**